### PR TITLE
Feature/combined header+sequence files

### PR DESCRIPTION
**Pull request by capnramses**
## Purpose of this PR

* Adds support for single-file volograms, Basis Universal textures in place of video texture file, integrated audio data, and the v1.3 vologram format.
* An example for loading single-file volograms, playback with Basis Universal textures, and extracting audio data.
* Volograms using the above tech can remove a dependency on vol_av, FFmpeg, and any other video players, which
    * Fixes sync issues with video players, particular with web players.
    * Reduces texture upload bandwidth to the GPU per-frame.
    * Is more platform-portable (both for licence portability, and build portability).
* Fix various security flaws and potential bugs and exploits.

## Changes made in this PR

image
New example 06 (`vol_geom_single_file.bin`) playing with Basis Universal textures.

* Updated Readme.md with new example.
* BasisU transcoder added under third_party/ subdirectory (please note this is a lot of files in this PR).
* Updated Makefile and windows build .bat for new example, and new vol_geom dependencies.
* Added new vol_basis library (wrapper of basisu transcoder).
* New example (06) program for playing single file volograms.
* Included a default vologram `samples/combined.vols` for combined, basis-texture files, with audio, in v1.3.
* Closed this security flaw in every function with no params: `a_function()` -> `a_function( void )`.
* Added an `is_file()` check when opening a file to avoid problems when passed a directory.

## Testing Summary

* Built and ran new example, as well as all old examples, on Ubuntu only.

### Build Command

* `make`
* `cd examples/`
* `./vol_geom_combined.bin` (or `.exe`).

### Not done in this PR

* WASM is not updated with new vol_geom.
* Web example using .basis files not added (coming soon in subsequent work).
* npm/cdn not updated.

## Risk Introduced

* Not tested on Windows on macOS.

# Pre-Merge Checklist

- [x] No conflicts with destination branch
- [x] Reviewed your own PR and checked your changed code
- [x] Invited reviewers (ideally +2)
- [x] All questions/comments answered (where necessary)
- [ ] Approvals received after latest changes (ideally 2+)
- [x] CI/CD builds successful

### CODE REVIEW

File: examples/06_vol_geom_single_file/main.c [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    compressed_texture.n_channels               = 4;
+    memset( &compressed_texture.properties, 0, sizeof( gfx_texture_properties_t ) );
+  }
+  gfx_mesh_t mesh = gfx_create_mesh_from_mem( NULL, 3, NULL, 2, NULL, 3, NULL, 0, 1, 0, true );
+  if ( !_update_mesh_with_frame( &mesh, 0, filename_vols, &vols_info, &compressed_texture ) ) {
+    fprintf( stderr, "update mesh with frame failed\n" );
+    return 1;
+  }
+
+  if ( vols_info.hdr.audio == 1 ) { // 0=no audio, 1=ogg vorbis/opus
```
Comment (ondrj006) [inline examples/06_vol_geom_single_file/main.c:116]: Should we change this to mp3 as this is what we are going to use?

File: src/vol_geom.h [hunk 2]
```diff
@@ -63,45 +64,41 @@ VOL_GEOM_EXPORT typedef struct vol_geom_short_str_t {
...
+  uint32_t topology;                // Removed in v1.3.
+  uint32_t frame_count;             //
+  bool normals;                     // Added in v1.2.
+  bool textured;                    // Added in v1.2.
+  uint8_t texture_compression;      // Added in v1.3.
+  uint8_t texture_container_format; // Added in v1.3.
+  uint32_t texture_width;           // Changed from int16_t to uint32_t in v1.3.
+  uint32_t texture_height;          // Changed from int16_t to uint32_t in v1.3.
+  float fps;                        // Added in v1.3. Most volograms are 30.0, but allows 29.97 and similar.
+  uint32_t audio;                   // Added in v1.3. { 0: no audio, 1: opus }.
```
Comment (ondrj006) [inline src/vol_geom.h:84]: we should redefine this 1: mp3

**Review by ondrj006 — APPROVED** [review]
I haven't find anything wrong. Only the audio is going to change in the future. We might want to use mp3 instead of ogg from the beginning?

**Review by patrick-volograms — APPROVED** [review]
Looks good. I have some wasm interface changes in the pipeline but I think this PR is big enough.

**Comment by patrick-volograms** [comment]
I'm getting some errors when I run `make` for testing:
```
thirdparty/basis_universal/transcoder/basisu.h:431:8: error: integral constant expression must have integral or unscoped enumeration type, not 'basisu::texture_format'
                case texture_format::cETC1:
                     ^~~~~~~~~~~~~~~~~~~~~
```

**Comment by capnramses** [comment]
> > I'm getting some errors when I run `make` for testing:
> > ```
> > thirdparty/basis_universal/transcoder/basisu.h:431:8: error: integral constant expression must have integral or unscoped enumeration type, not 'basisu::texture_format'
> >                 case texture_format::cETC1:
> >                      ^~~~~~~~~~~~~~~~~~~~~
> > ```
> 
> I fixed the error by changing the CPP standard in the Makefile `CPP = clang++ -std=c++11`

Thanks! Change made.

File: examples/06_vol_geom_single_file/main.c (hunk 1 above)
Comment (capnramses) [reply examples/06_vol_geom_single_file/main.c]: thanks. just changed to a generic statement for now. i would like to make the demo play audio but can move that into a subsequent PR to break it up into smaller pieces.

File: src/vol_geom.h (hunk 2 above)
Comment (capnramses) [reply src/vol_geom.h]: just removed the format-specific comment for this PR

**Review by ondrj006 — APPROVED** [review]

**Pull request merged**